File: core/src/main/scala/com/xhachi/gae4s/datastore/meta/property.java
package com.xhachi.gae4s.datastore.meta;

import java.lang.annotation.*;

@Inherited
@Retention(RetentionPolicy.CLASS)
@Target({ElementType.FIELD, ElementType.METHOD})
public @interface property {

    String name() default "";
    boolean json() default false;
    boolean serialize() default false;
    boolean indexed() default false;
    boolean version() default false;
    boolean creationDate() default false;
    boolean modificationDate() default false;
    int order() default Integer.MAX_VALUE;

}

File: gae4s-core/src/main/scala/com/xhachi/gae4s/datastore/Property.scala
package com.xhachi.gae4s.datastore

sealed trait Property[T] {
  def name: String

  def value: T

  def withValue(value: T): this.type
}

case class UnindexedProperty[T](name: String, value: T) extends Property[T] {
  def withValue(value: T): this.type = UnindexedProperty[T](name, value).asInstanceOf[this.type]
}

case class IndexedProperty[T](name: String, value: T) extends Property[T] {
  def withValue(value: T): this.type = IndexedProperty[T](name, value).asInstanceOf[this.type]
}

case class VersionProperty(name: String, value: Long) extends Property[Long] {
  def withValue(value: Long): this.type = VersionProperty(name, value).asInstanceOf[this.type]
}
